### ISSUE TITLE
cbn without zeta substitutes let-bindings

### ISSUE THREAD

**Issue by ichung** [coq/coq#8270]
Version 8.8.1.

`cbn`, when called without the zeta reduction flag, performs a sort of intermediate reduction on let-bindings where it substitutes the variable in the body of the let, without eliminating the let-binding.  This is a bug because without zeta, `cbn` shouldn't change the lets at all.  This bug makes `cbn` unsuitable for performing reduction when using let-bindings to avoid an exponential blowup in term size, which is inconvenient when one wants to simplify applications of fixpoints to such expressions.

Related to #5327 along the general theme that it's inconvenient to perform reduction in Coq without unfolding let-bindings.

```
Goal let x := 0 in
     let y := x in
     y = 0.
  (* We use "cofix" as an example because there are obviously no
     cofixpoints in sight.  This problem arises with any set of
     reduction flags (not including zeta where the lets are of course reduced away) *)
  cbv cofix.
  (* as expected, the goal is unchanged. *)
  Set Debug RAKAM.
  cbn cofix.
  (* cbn substitutes the bodies of the let-expressions without
     eliminating them.  This is different from zeta-reduction (which
     would eliminate the let-bindings entirely) and is a bug; cbn ought
     to leave the goal unchanged just like cbv. *)
```

[Output from Debug RAKAM](https://gist.github.com/ichung/0e2f84912ca961927a9a8108656ed1cf)

**Comment by gares** [comment]
I think @herbelin is the most knowledgeable about the reduction machine used by cbn.

**Comment by herbelin** [comment]
Fixed in #8285.
